src/main/scala/xiangshan/backend/issue/Dispatch2Iq.scala

```diff
@@ -139,6 +142,352 @@ abstract class Dispatch2IqImp(override val wrapper: Dispatch2Iq)(implicit p: Par
+  val IQ0Enq1Select2 = !IQ0Enq1Select1 && !IQ0Enq0Select2 && uopsInDq0(2).valid
+  val IQ1Enq0Select2 = !IQ1Enq0Select0 && !IQ1Enq0Select1 && !IQ0Enq0Select2 && !IQ0Enq1Select2 && uopsInDq0(2).valid
+  val IQ1Enq1Select2 = !IQ0Enq0Select2 && !IQ0Enq1Select2 && !IQ1Enq0Select2 && uopsInDq0(2).valid
+  val IQ0Enq1Select3 = !IQ0Enq1Select1 && !IQ0Enq1Select2 && uopsInDq0(3).valid
+  val IQ1Enq1Select3 = !IQ1Enq1Select1 && !IQ1Enq1Select2 && uopsInDq0(3).valid
+  when(uopsOutDq0.head.ready && uopsOutDq0.last.ready){
+    IQ0Enq0Select := Cat(false.B, IQ0Enq0Select2, IQ0Enq0Select1, IQ0Enq0Select0).asBools
+    IQ0Enq1Select := Cat(IQ0Enq1Select3, IQ0Enq1Select2, IQ0Enq1Select1, false.B).asBools
+    IQ1Enq0Select := Cat(false.B, IQ1Enq0Select2, IQ1Enq0Select1, IQ1Enq0Select0).asBools
+    IQ1Enq1Select := Cat(IQ1Enq1Select3, IQ1Enq1Select2, IQ1Enq1Select1, false.B).asBools
+  }.elsewhen(uopsOutDq0.head.ready){
+    IQ0Enq0Select := Cat(false.B, false.B, false.B, true.B).asBools
+    IQ0Enq1Select := Cat(false.B, false.B, true.B, false.B).asBools
+    IQ1Enq0Select := Cat(false.B, false.B, false.B, false.B).asBools
+    IQ1Enq1Select := Cat(false.B, false.B, false.B, false.B).asBools
+  }.elsewhen(uopsOutDq0.last.ready){
+    IQ0Enq0Select := Cat(false.B, false.B, false.B, false.B).asBools
+    IQ0Enq1Select := Cat(false.B, false.B, false.B, false.B).asBools
+    IQ1Enq0Select := Cat(false.B, false.B, false.B, true.B).asBools
+    IQ1Enq1Select := Cat(false.B, false.B, true.B, false.B).asBools
+  }.otherwise{
+    IQ0Enq0Select := Cat(false.B, false.B, false.B, false.B).asBools
+    IQ0Enq1Select := Cat(false.B, false.B, false.B, false.B).asBools
+    IQ1Enq0Select := Cat(false.B, false.B, false.B, false.B).asBools
+    IQ1Enq1Select := Cat(false.B, false.B, false.B, false.B).asBools
+  }
+  assert(PopCount(IQ0Enq0Select) < 2.U, "IQ0Enq0Select is not one hot")
+  assert(PopCount(IQ0Enq1Select) < 2.U, "IQ0Enq1Select is not one hot")
+  assert(PopCount(IQ1Enq0Select) < 2.U, "IQ1Enq0Select is not one hot")
+  assert(PopCount(IQ1Enq1Select) < 2.U, "IQ1Enq1Select is not one hot")
+  assert(PopCount(IQ0Enq0Select) < 2.U, "IQ0Enq0Select is not one hot")
+  assert(PopCount(IQ0Enq1Select) < 2.U, "IQ0Enq1Select is not one hot")
+  assert(PopCount(IQ1Enq0Select) < 2.U, "IQ1Enq0Select is not one hot")
+  assert(PopCount(IQ1Enq1Select) < 2.U, "IQ1Enq1Select is not one hot")
+  assert(PopCount(VecInit(IQ0Enq0Select(0), IQ0Enq1Select(0), IQ1Enq0Select(0), IQ1Enq1Select(0))) < 2.U, "Dq1Uop0Select is not one hot")
+  assert(PopCount(VecInit(IQ0Enq0Select(1), IQ0Enq1Select(1), IQ1Enq0Select(1), IQ1Enq1Select(1))) < 2.U, "Dq1Uop1Select is not one hot")
+  assert(PopCount(VecInit(IQ0Enq0Select(2), IQ0Enq1Select(2), IQ1Enq0Select(2), IQ1Enq1Select(2))) < 2.U, "Dq1Uop2Select is not one hot")
+  assert(PopCount(VecInit(IQ0Enq0Select(3), IQ0Enq1Select(3), IQ1Enq0Select(3), IQ1Enq1Select(3))) < 2.U, "Dq1Uop3Select is not one hot")
+  val dq0Block = Wire(Vec(4, Bool()))
```

Comment:
PLZ explain how to get these immediate numbers and name them.

Reply:
Modified